Dataset.py

```diff
@@ -158,6 +158,12 @@ def __init__(self, name=None,
       context_window = NumbersDict(broadcast_value=0, numbers_dict={"data": context_window})
     elif isinstance(context_window, dict):
       context_window = NumbersDict(broadcast_value=0, numbers_dict=context_window)
+    elif isinstance(context_window, (tuple, list)):
+      assert len(context_window) == 2
+      # assume that first element of tuple|list is for left context and second element for right context
+      self.ctx_left = NumbersDict(broadcast_value=0, numbers_dict=context_window[0])
+      self.ctx_right = NumbersDict(broadcast_value=0, numbers_dict=context_window[1])
```

Comment:
Never conditionally assign attributes. That very easily can lead to unexpected behavior and bugs.
Instead, move this case (`instance .. tuple`) to the top, as the first `if`, then do this logic, and then put a single `else:` branch, where you put all the other `context_window` logic, and instead of assigning `self.context_window` there, you also calculate `ctx_left`/`ctx_right` (copy the logic from `_get_context_window_left_right` here), and so you only have these attributes, and no `self.context_window` at all anymore.